dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/programrule/engine/RuleActionScheduleMessageImplementer.java

```diff
@@ -0,0 +1,151 @@
+import org.hisp.dhis.program.ProgramInstance;
+import org.hisp.dhis.program.ProgramStageInstance;
+import org.hisp.dhis.program.notification.ProgramNotificationInstance;
+import org.hisp.dhis.program.notification.ProgramNotificationTemplate;
+import org.hisp.dhis.rules.models.RuleAction;
+import org.hisp.dhis.rules.models.RuleActionScheduleMessage;
+import org.hisp.dhis.rules.models.RuleEffect;
+import org.hisp.dhis.system.util.DateUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+
+/**
+ * @Author Zubair Asghar.
+ */
+public class RuleActionScheduleMessageImplementer extends BaseRuleActionImplementer
+{
+    // -------------------------------------------------------------------------
+    // Dependencies
+    // -------------------------------------------------------------------------
+
+    @Autowired
+    private NotificationLoggingService notificationLoggingService;
+
+    @Autowired @Qualifier( "org.hisp.dhis.program.notification.ProgramNotificationInstanceStore" )
+    private IdentifiableObjectStore<ProgramNotificationInstance> programNotificationInstanceStore;
+
+    @Override
+    public boolean accept( RuleAction ruleAction )
+    {
+        return ruleAction != null && ruleAction instanceof RuleActionScheduleMessage;
+    }
+
+    @Override
+    public void implement( RuleEffect ruleEffect, ProgramInstance programInstance )
+    {
+        if ( !validate( ruleEffect, programInstance ) )
+        {
+            return;
+        }
+
+        ProgramNotificationTemplate template = getNotificationTemplate( ruleEffect.ruleAction() );
+
+        String key = generateKey( template, programInstance );
+
+        String date = ruleEffect.data();
+
+        if ( !isDateValid( date ) )
+        {
+            return;
+        }
+
+        ProgramNotificationInstance notificationInstance = createNotificationInstance( template, date );
+        notificationInstance.setProgramStageInstance( null );
+        notificationInstance.setProgramInstance( programInstance );
+
+        programNotificationInstanceStore.save( notificationInstance );
+
+        ExternalNotificationLogEntry entry = createLogEntry( key, template.getUid() );
+        entry.setNotificationTriggeredBy( NotificationTriggerEvent.PROGRAM );
+        notificationLoggingService.save( entry );
+    }
+
+    @Override
+    public void implement( RuleEffect ruleEffect, ProgramStageInstance programStageInstance )
+    {
+        if ( !validate( ruleEffect, programStageInstance.getProgramInstance() ) )
+        {
+            return;
+        }
+
+        ProgramNotificationTemplate template = getNotificationTemplate( ruleEffect.ruleAction() );
+
+        String key = generateKey( template, programStageInstance.getProgramInstance() );
+
+        String date = ruleEffect.data();
+
+        if ( !isDateValid( date ) )
+        {
+            return;
+        }
```

Comment:
Same comment as above, it seems to potentially be more that can be centralized in the abstract base class.